File: apps/mobile/lib/core/models/pharmacy.dart
/// Pharmacy model
class Pharmacy {
  final String id;
  final String name;
  final String address;
  final String city;
  final String district;
  final String phone;
  final double latitude;
  final double longitude;
  final bool isOnDuty;
  final double? distanceKm;

  const Pharmacy({
    required this.id,
    required this.name,
    required this.address,
    required this.city,
    required this.district,
    required this.phone,
    required this.latitude,
    required this.longitude,
    this.isOnDuty = true,
    this.distanceKm,
  });

  factory Pharmacy.fromJson(Map<String, dynamic> json) {
    return Pharmacy(
      id: json['id'] as String,
      name: json['name'] as String,
      address: json['address'] as String,
      city: json['city'] as String,
      district: json['district'] as String,
      phone: json['phone'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      isOnDuty: json['is_on_duty'] as bool? ?? true,
      distanceKm: json['distance_km'] != null
          ? (json['distance_km'] as num).toDouble()
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'address': address,
      'city': city,
      'district': district,
      'phone': phone,
      'latitude': latitude,
      'longitude': longitude,
      'is_on_duty': isOnDuty,
      'distance_km': distanceKm,
    };
  }

  /// Format distance for display
  String get formattedDistance {
    if (distanceKm == null) return '';
    if (distanceKm! < 1) {
      return '${(distanceKm! * 1000).round()} m';
    }
    return '${distanceKm!.toStringAsFixed(1)} km';
  }

  Pharmacy copyWith({
    String? id,
    String? name,
    String? address,
    String? city,
    String? district,
    String? phone,
    double? latitude,
    double? longitude,
    bool? isOnDuty,
    double? distanceKm,
  }) {
    return Pharmacy(
      id: id ?? this.id,
      name: name ?? this.name,
      address: address ?? this.address,
      city: city ?? this.city,
      district: district ?? this.district,
      phone: phone ?? this.phone,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      isOnDuty: isOnDuty ?? this.isOnDuty,
      distanceKm: distanceKm ?? this.distanceKm,
    );
  }
}

File: services/api/app/routes/pharmacy.py
"""Pharmacy routes."""

from fastapi import APIRouter, Query, HTTPException
from pydantic import BaseModel
from typing import Optional

from app.providers.mock_provider import MockPharmacyProvider
from app.core.cache import cache

router = APIRouter()

# Provider instance (will be configurable via env)
provider = MockPharmacyProvider()


class Pharmacy(BaseModel):
    """Pharmacy model."""

    id: str
    name: str
    address: str
    city: str
    district: str
    phone: str
    latitude: float
    longitude: float
    is_on_duty: bool = True
    distance_km: Optional[float] = None


class PharmacyResponse(BaseModel):
    """Response model for pharmacy list."""

    pharmacies: list[Pharmacy]
    total: int
    cached: bool = False


@router.get("/on-duty", response_model=PharmacyResponse)
async def get_on_duty_pharmacies(
    city: Optional[str] = Query(None, description="Şehir adı"),
    district: Optional[str] = Query(None, description="İlçe adı"),
    lat: Optional[float] = Query(None, description="Enlem"),
    lon: Optional[float] = Query(None, description="Boylam"),
):
    """
    Nöbetçi eczaneleri getir.

    - city/district ile arama
    - lat/lon ile yakınlık bazlı arama
    """
    # Validate input
    if not city and not (lat and lon):
        raise HTTPException(
            status_code=400,
            detail="Şehir veya konum (lat/lon) belirtmelisiniz",
        )

    # Check cache
    cache_key = f"pharmacies:{city}:{district}:{lat}:{lon}"
    cached_data = cache.get(cache_key)
    if cached_data:
        return PharmacyResponse(
            pharmacies=cached_data,
            total=len(cached_data),
            cached=True,
        )

    # Fetch from provider
    if lat and lon:
        pharmacies = await provider.get_by_location(lat, lon)
    else:
        pharmacies = await provider.get_by_city(city, district)

    # Cache result
    cache.set(cache_key, pharmacies)

    return PharmacyResponse(
        pharmacies=pharmacies,
        total=len(pharmacies),
        cached=False,
    )
